tools/oesign/oesign.c

```diff
@@ -378,6 +467,24 @@ void _merge_config_file_options(
     if (options->security_version.has_value)
         properties->config.security_version = options->security_version.value;
 
+    bool kss_supported = _is_kss_supported();
```

Comment:
With regard to Intel SDM vol. 3d description of EINIT, I think the documentation missed the ISVEXTPRODID check. Can someone from Intel confirm?

I think many developers won't be fully aware that using non-zero ISVFAMILYID/ISVEXTPRODID implies the SGX enclave will only run on SGX CPU with KSS support. But I think the tool chain (oesign and/or the loader) can and should own SIGSTRUCT.ATTRIBUTES.KSS, SIGSTRUCT.ATTRIBUTEMASK.KSS and SECS.ATTRIBUTE.KSS setting on behalf of the developer.

SGX CPU with KSS support enforces the following policy:

- If (SIGSTRUCT.ATTRIBUTEMASK.KSS ==0), loader can set SECS.ATTRIBUTES.KSS as 1 or 0. 

- If (SIGSTRUCT.ATTRIBUTES.KSS ==1 AND SIGSTRUCT.ATTRIBUTEMASK.KSS ==1), loader must set SECS.ATTRIBUTES.KSS as 1.

- If (SIGSTRUCT.ATTRIBUTES.KSS ==0 AND SIGSTRUCT.ATTRIBUTEMASK.KSS ==1), loader must set SECS.ATTRIBUTES.KSS as 0.  

- If SIGSTRUCT contains non-zero ISVFAMILYID/ISVEXTPRODID, loader must set SECS.ATTRIBUTES.KSS as 1

- If loader sets SECS with non-zero CONFIGID/CONFIGSVN, loader must also set SECS.ATTRIBUTES.KSS as 1

SGX CPU without KSS support enforces the following policy:

- SIGSTRUCT.ATTRIBUTES.KSS, SIGSTRUCT.ATTRIBUTEMASK.KSS must be 0

- SIGSTRUCT.ISVFAMILYID, SIGSTRUCT.ISVEXTPRODID must be 0

- SECS.ATTRIBUTES.KSS, SECS.CONFIGID, SECS.CONFIGSVN must be 0

OE loader should enumerates SGX CPU support for KSS, and set SECS according to the CPU capability, SIGSTRUCT and to not violate the CPU enforced policy listed above. With regard to SECS.ATTRIBUTES.KSS setting (on CPU with KSS support) the loader can follow a simple logic:

- Set SECS.ATTRIBUTES.KSS = 1, if the the CPU supports KSS, unless SIGSTRUCT. ATTRIBUTES.KSS=0 and SIGSTRUCT.ATTRIBUTEMASK.KSS=1.

If the developer does not intend to use any KSS related feature, the loader enabling KSS (on CPU with KSS support) and setting SECS.CONFIGID, SECS.CONFIGSVN as 0 does not violate any function or security policy of the enclave. (
The proposed [OE_ENCLAVE_SETTING_CONTEXT_PRE_MEASURED_ID data for CONFIGID/CONFIGSVN](https://github.com/openenclave/openenclave/pull/3269) decide what values to use for SECS.CONFIGID, SECS.CONFIGSVN when the developer intends to use the CONFIGID/CONFIGSVN feature of KSS. But it's out of scope of this PR).

Oesign should set SIGSTRUCT.ATTRIBUTES.KSS =0 and SIGSTRUCT.ATTRIBUTEMASK.KSS =0 by default, unless the developer sets non-zero ISVFAMILYID/ISVEXTPRODID. In that case, oesign sets SIGSTRUCT.ATTRIBUTES.KSS =1 and SIGSTRUCT.ATTRIBUTEMASK.KSS =1. The signing process should also allow the developer to use advanced interface to set SIGSTRUCT.ATTRIBUTES.KSS =0 and SIGSTRUCT.ATTRIBUTEMASK.KSS =1.

With SIGSTRUCT.ATTRIBUTES.KSS =0 and SIGSTRUCT.ATTRIBUTEMASK.KSS =0, developers not aware of KSS will be building enclaves that run on both SGX CPUs with and without KSS support.

Reply:
Thanks for the write-up @bodzhang! To summarize my understanding, the difference from the previous 3 scenarios I laid out is that opting into KSS for ISVFAMILYID/ISVEXTPRODID constrains what the loader can do at runtime, and the latter can be implemented on top of of this. To phase this work out, I would suggest an ordering of implementation based on the semantics you've defined:

1. If a developer/signer uses ISVFAMILYID/ISVEXTPRODID, resulting enclave should have `SIGSTRUCT.ATTRIBUTES.KSS=1` and `SIGSTRUCT.ATTRIBUTEMASK.KSS=1`
   - This should be implemented as part of this PR, and should include both `OE_SET_ENCLAVE_SGX` and oesign codepaths.
   - OE SDK uses hardcoded `SGX_SIGSTRUCT_ATTRIBUTEMASK_FLAGS` today, where `SIGSTRUCT.ATTRIBUTEMASK.KSS=1` is true by default, so the PR only has to deal with setting `SIGSTRUCT.ATTRIBUTES.KSS=1` or not.
   - This PR needs to add an update to `OE_MAKE_ATTRIBUTES` or an equivalent implmentation for `OE_SET_ENCLAVE_SGX` to be consistent with the oesign.

2. By default, if a developer/signer does not specify ISVFAMILYID/ISVEXTPRODID, resulting enclave should have `SIGSTRUCT.ATTRIBUTES.KSS=0` and `SIGSTRUCT.ATTRIBUTEMASK.KSS=1`
   - Note that I have inverted the default proposed by Bo here, and I'm suggesting that we should not change the default in this PR.
   - The existing SDK still does not support CONFIGID/CONFIGSVN, so it should be treated as an unintended change in identity if the enclave is loaded by some loader that permits setting those properties at runtime.
   - The `SIGSTRUCT.ATTRIBUTES.KSS=0` and `SIGSTRUCT.ATTRIBUTEMASK.KSS=1` combination specifies intent of blocking that unintended scenario, and represents a safe default consistent with older versions of the SDK until CONFIGID/CONFIGSVN is supported. At that point, we can also revisit changing the default KSS attribute mask as a breaking change.

3. The developer/signer can specify that they do not care if the enclave is loaded with CONFIGID/CONFIGSVN and the resulting enclave should have `SIGSTRUCT.ATTRIBUTES.KSS=0` and `SIGSTRUCT.ATTRIBUTEMASK.KSS=0`.
   - Whether this becomes the default or not, it is an additional option we need to expose to a developer who doesn't use ISVFAMILYID/ISVEXTPRODID but may want their enclave to run with CONFIGID/CONFIGSVN.
   - I am in favor of making this an explicit opt-in because it is not obvious to a developer that their enclave may run with additional CONFIGID/CONFIGSVN properties.
      - On the flip side, this doesn't appear to be immediately harmful to the enclave since any attestation system that doesn't expect to use the CONFIGID/CONFIGSVN won't interact with it anyway. It may be undesirable for other reasons as it allows the host to inject data into the report which might be used for packet tagging perhaps.
   - In terms of code change, this requires changing the existing hardcoded `SGX_SIGSTRUCT_ATTRIBUTEMASK_FLAGS` so that the SIGSTRUCT.ATTRIBUTEMASK can be set dynamically by `OE_SET_ENCLAVE_SGX` and oesign.

4. There's an additional narrow case of allowing a developer/signer to specify that enclave _must_ run with CONFIGID/CONFIGSVN by specifying `SIGSTRUCT.ATTRIBUTES.KSS=1` and `SIGSTRUCT.ATTRIBUTEMASK.KSS=1` when ISVFAMILYID/ISVEXTPRODID is not provided. 
   - I think there's some additional design discussion for when CONFIGID/CONFIGSVN is added as to whether we need to support this. Off the top of my head, I can't think of a really strong reason to require this, but it's also not a lot of work. The real question is explaining to a dev why they would want to do this.
   - If we do support this case, `oe_create_enclave` should provide check and provide a specific error trace in this configuration and it detects none of the KSS properties are defined. 

5. For completeness, there should not be a case that `OE_SET_ENCLAVE_SGX` or oesign needs to support for setting `SIGSTRUCT.ATTRIBUTES.KSS=1` and `SIGSTRUCT.ATTRIBUTEMASK.KSS=0` at development/sign time.
   - This basically means "I want the loader to use KSS by default, but it can choose not to" and is no different in practical terms from setting `SIGSTRUCT.ATTRIBUTES.KSS=0` and `SIGSTRUCT.ATTRIBUTEMASK.KSS=0`, so we should standardize on that encoding instead.
      - The loader doesn't need `SIGSTRUCT.ATTRIBUTES.KSS=1` for it to set `SECS.ATTRIBUTES.KSS=1` in that case.
      - If the dev/signer uses `ISVFAMILYID/ISVEXTPRODID`, it becomes case (1) as it is not legal to set `SIGSTRUCT.ATTRIBUTEMASK.KSS=0`